### PR TITLE
Add third RetryResult (instead of using null) to denote 'proceed'

### LINKED ISSUE #105: ExceptionHandler.Interceptor.RetryResult should have a third value, used instead of null
The methods of [ExceptionHandler.Interceptor](http://googlecloudplatform.github.io/gcloud-java/apidocs/com/google/gcloud/ExceptionHandler.Interceptor.html) return a value that is one of `RetryResult.RETRY`, `RetryResult.ABORT`, or `null`. Instead of `null`, there should be a third `RetryResult` constant. Returning null from the `beforeEval` or `afterEval` methods would then provoke a `NullPointerException`.

### PR DESCRIPTION
**Pull request by ajkannan**
This PR addresses fixes #105

### CODE REVIEW

File: gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java [hunk 1]
```diff
@@ -253,18 +241,22 @@ void verifyCaller(Callable<?> callable) {
 
   boolean shouldRetry(Exception ex) {
     for (Interceptor interceptor : interceptors) {
-      Interceptor.RetryResult retryResult = interceptor.beforeEval(ex);
-      if (retryResult != null) {
-        return retryResult.booleanValue();
+      Interceptor.RetryResult retryResult = checkNotNull(interceptor.beforeEval(ex));
+      if (retryResult != Interceptor.RetryResult.PROCEED) {
+        return interceptor.beforeEval(ex) == Interceptor.RetryResult.RETRY;
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java]: Why do you call the interceptor twice?

Comment (ajkannan) [inline gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java]: Noted, I'll remove the second call.

File: gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java [hunk 2]
```diff
@@ -48,18 +47,7 @@
...
-      private final boolean booleanValue;
-
-      RetryResult(boolean booleanValue) {
-        this.booleanValue = booleanValue;
-      }
-
-      boolean booleanValue() {
-        return booleanValue;
-      }
+      ABORT, RETRY, PROCEED;
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java]: Can we make the enum values more clear (and the javadoc).
Maybe `NO_RETRY`, `RETRY` and `PROCEED_EVALUATION` (maybe you can think of a better name for the last one).

File: gcloud-java-core/src/test/java/com/google/gcloud/ExceptionHandlerTest.java [hunk 3]
```diff
@@ -158,11 +158,60 @@ public RetryResult beforeEval(Exception exception) {
...
+      public RetryResult afterEval(Exception exception, RetryResult retryResult) {
+        return null;
+      }
+
+    };
+
+    ExceptionHandler handler1 = ExceptionHandler.builder().interceptor(interceptor1).build();
+    try {
+      handler1.shouldRetry(new Exception());
+      fail("Expected null pointer exception due to null RetryResult from beforeEval");
```
Comment (aozarov) [inline gcloud-java-core/src/test/java/com/google/gcloud/ExceptionHandlerTest.java]: Instead you can use @Rule. see [here](http://junit.org/apidocs/org/junit/rules/ExpectedException.html).

File: gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java (hunk 2 above)
Comment (ajkannan) [inline gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java]: I'll change the names to `NO_RETRY`, `RETRY`, and `CONTINUE_EVALUATION`.

**Comment by ajkannan** [comment]
Related question, but perhaps not for this PR: should we write tests for DatastoreImpl.java and StorageImpl.java to make sure the interceptor works as intended?

**Comment by aozarov** [comment]
> Related question, but perhaps not for this PR: should we write tests for DatastoreImpl.java and StorageImpl.java to make sure the interceptor works as intended?

Yes, I think we should test exception handling in the Datastore. 
Can you create a separate issue for it?

**Pull request merged**